rfcs/0049-temporal-join.md

```diff
@@ -0,0 +1,103 @@
+
+ // Output
+ EventTimeTemporalJoin -> WaterMark[8] -> JoinedRow(+Row[7], Null) -> JoinedRow(+Row[5], +Row[2]) -> WaterMark[1]
+```
+
+![event_time_temporal_join](images/0049-temporal-join/event_time_temporal_join.svg)
+
+### State Cleaning
+
+The outer side will clean up its state's outdated data (which is behind current watermark). The inner side will clean up its outdated data too, but keep at least one row for each primary key. **We need to decide the state's primary key ordering : the join key first and then the watermark key or the watermark key first and then the join key**. 
+
+#### The Outer Side
+
+For the outer side, it is better to choose the watermark key first and then the join key as the primary key ordering, because each time the join watermark increases, we need to join all rows between the previous watermark and current watermark. Moreover, using this ordering is helpful to search and clean up the outer side's state.
+
+
+#### The Inner Side
+
+If the watermark between both sides will not vary too much so that we can clean up the state quickly and unprocessed input messages will be small enough to be kept in the memory buffer. Use **the watermark key first and then the join key** as the state primary key is better.
+
+As we need to keep at least one row for each primary key, it is more appropriate to maintain two states: the main table and the history table. The main table always stores the latest version of records and its primary key is the join key. The history table stores the other versions of records and its primary key is in the order of the watermark key first and then the join key. In this way we can clean up the state quickly (only scan the history table) and keep at least one row for each primary key.
+
+Otherwise, if there is too much data on the inner side, Use **the join key first and then the watermark key** as the state primary key is better. In this way, we can just maintain one state and use the memory as a cache. A state cleaning is performed along with access to the inner side table at the same time.
+
+Considering how to handle watermarks for `Mv on Mv`, it needs to consume all the snapshot data of the upstream mv before producing watermarks. The data size could be huge, so using the join key first and then the watermark key as the state primary key may be better.
+
+
+## Summary
+
+- Temporal join requires the outer side to be append-only and the inner side's primary key contained in the equivalence condition of the temporal join condition
+- Process time temporal join can be stateless. It relies on the scheduler to place the join operator together with the inner side table.
+- Event time temporal join requires watermark columns and maintaining states for both input sides. The outer side state's primary key should be in the order of the watermark key first and then the join key, while the inner side state's primary key should be in the order of join key first and then watermark key
```

Comment:
Correct me if I'm wrong, but according to the description, it seems that event time temporal join is the same as a regular join with two watermark-filtered inputs which emits on the watermark close? 🤔

Reply:
Watermark-filtered input will delete rows as watermark increases, but event time temporal join needs to keep at least one row for each primary key of the inner side table.